### PR TITLE
close #938 performance enhancement for watch

### LINKED ISSUE #938: Improve watch performance
I've noticed that when we use `watch('field')` it gets the values for all fields in the form before it can actually watch for it.
This is problematic when using `useFieldArray` which could potentially have lots of rows that can be watched.

image

here is what the current code looks like: 
```
    function watch(fieldNames, defaultValue) {
        const combinedDefaultValues = isUndefined(defaultValue)
            ? isUndefined(defaultValuesRef.current)
                ? {}
                : defaultValuesRef.current
            : defaultValue;
        // get values of every single field in the form.
        const fieldValues = getFieldsValues(fieldsRef.current);
```
one solution could be to only get the values of the fields that needs to watched.
```
        const fieldValues = getFieldsValues(fieldsRef.current, fieldNames);
```
and then update `getFieldsValues` to something like:

```
var getFieldsValues = (fields, filter) => {
    const values = {};
    Object.values(fields).forEach(({ ref, ref: { name } }) => {
        if (!filter || filter.indexOf(name) > -1) {
            values[name] = getFieldValue(fields, ref);
        }
    });
    return values;
}
```
I'm not sure what other things this could impact but I've seem rendering time dropping from 114ms to 42ms with this change.

### PR DESCRIPTION
**Pull request by bluebill1049**

### CODE REVIEW

**Comment by bluebill1049** [comment]
@kotarella1110 if you have sometimes, please help on improve TS :)

**Comment by kotarella1110** [comment]
I'll take a look later!

File: src/logic/getFieldsValues.ts [hunk 1]
```diff
@@ -1,13 +1,19 @@
...
-  (Object.values(fields) as Field[]).reduce(
-    (previous, { ref, ref: { name } }) => ({
-      ...previous,
-      ...{ [name]: getFieldValue(fields, ref) },
-    }),
-    {} as FormValues,
-  );
+  search?: any,
+) => {
+  const output: Record<string, FormValues> = {};
```
Comment (kotarella1110) [inline src/logic/getFieldsValues.ts]: how is this?

```ts
const output: Partial<Record<FieldName<FormValues>, FormValues>> = {};
```

File: src/logic/getFieldsValues.ts [hunk 2]
```diff
@@ -1,13 +1,19 @@
...
+  const isSearchString = isString(search);
+
+  for (const name in fields) {
+    if (!isSearchString || (isSearchString && name.startsWith(search))) {
+      output[name] = getFieldValue(fields, (fields[name] || {}).ref);
```
Comment (kotarella1110) [inline src/logic/getFieldsValues.ts]:
```ts
output[name as FieldName<FormValues>] = getFieldValue(fields, (fields[name] || {}).ref);
```

**Comment by kotarella1110** [comment]
By the way, the `FieldName` type is effectively a `string` type.

```ts
export type FieldName<FormValues extends FieldValues> =
  | (keyof FormValues & string)
  | string
// => string
```

Is this a problem if fixed as follows?

```ts
export type FieldName = string
```

**Comment by bluebill1049** [comment]
> By the way, the `FieldName` type is effectively a `string` type.
> 
> ```ts
> export type FieldName<FormValues extends FieldValues> =
>   | (keyof FormValues & string)
>   | string
> // => string
> ```
> 
> Is this a problem if fixed as follows?
> 
> ```ts
> export type FieldName = string
> ```

This is more for a reference point of view.

File: src/logic/getFieldsValues.ts (hunk 1 above)
Comment (bluebill1049) [reply src/logic/getFieldsValues.ts]: sounds good :)

File: src/logic/getFieldsValues.ts (hunk 2 above)
Comment (bluebill1049) [reply src/logic/getFieldsValues.ts]: sounds good :)

**Comment by bluebill1049** [comment]
please commit your suggested change :)

**Comment by kotarella1110** [comment]
Changing the `FieldName` type to `string` simplifies the type.

`type FieldName<FormValues extends FieldValues> = (keyof FormValues & string) | string`:

スクリーンショット 2020-01-31 16 34 39

`type FieldName = string`:

スクリーンショット 2020-01-31 16 35 27

**Comment by bluebill1049** [comment]
> Changing the `FieldName` type to `string` simplifies the type.
> 
> `type FieldName<FormValues extends FieldValues> = (keyof FormValues & string) | string`:
> 
> スクリーンショット 2020-01-31 16 34 39
> 
> `type FieldName = string`:
> 
> スクリーンショット 2020-01-31 16 35 27

ok if that's the case, let's go with it then.

**Review by kotarella1110 — APPROVED** [review]

**Comment by bluebill1049** [comment]
@kotarella1110 you want to make the change to `string`? then we can merge this PR

**Comment by bluebill1049** [comment]
actually let's merge this PR, you can create a separate PR.

**Comment by kotarella1110** [comment]
@bluebill1049 I want to it! But changing the `FieldName` type requires us to fix it in large scope.
Let's merge this once. After that, I create a new PR.

**Comment by bluebill1049** [comment]
yea take your time. love to have some TS upgrade from you ❤️

**Pull request merged**